### PR TITLE
Hiding Feign annotations using a Mocca InvocationHandlerFactory #40

### PR DESCRIPTION
**Pull request by fabiocarvalho777**

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -140,7 +140,10 @@ subprojects {
     }
 }
 
-def feign_version = '11.5'
+def feign_version = '11.7'
```
Comment (jamesmartinpp) [inline build.gradle]: Perhaps we could do the upgrade in a separate PR? Not a big deal though.

File: mocca-client/src/main/java/com/paypal/mocca/client/MoccaExceptionHandler.java [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+    }
+
+    static RuntimeException handleException(Throwable throwable) {
+        if (throwable instanceof FeignException) {
+            Throwable cause = throwable.getCause();
+            if (acceptableExceptions.contains(cause.getClass()) && cause instanceof RuntimeException) {
+                return (RuntimeException) cause;
+            }
+        }
+        return new MoccaException("An exception has happened when invoking the client target", throwable);
```
Comment (jamesmartinpp) [inline mocca-client/src/main/java/com/paypal/mocca/client/MoccaExceptionHandler.java]:
```suggestion
        return new MoccaException("The invocation of a client method has resulted in an exception: ", throwable);
```

File: mocca-client/src/main/java/com/paypal/mocca/client/MoccaFeignEncoder.java [hunk 3]
```diff
@@ -99,9 +99,7 @@ void validateVariables(Object[] parameters, RequestTemplate template) {
         Set<ConstraintViolation<Object>> violationSet = validator.forExecutables().validateParameters(client, method, parameters);
 
         if (violationSet.size() > 0) {
-            throw new EncodeException("Constraint violations found in one or more request parameters",
-                    new ConstraintViolationException(violationSet)
-            );
+            throw new ConstraintViolationException(violationSet);
```
Comment (jamesmartinpp) [inline mocca-client/src/main/java/com/paypal/mocca/client/MoccaFeignEncoder.java:102]: Is it clear that a constraint violation has occurred without the message? Just wondering.

File: mocca-client/src/main/java/com/paypal/mocca/client/MoccaFeignInvocationHandlerFactory.java [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+ * when using a custom Feign invocation handler factory is possible
+ *
+ * @author facarvalho, crankydillo@gmail.com
+ */
+class MoccaFeignInvocationHandlerFactory implements InvocationHandlerFactory {
+
+    private final InvocationHandlerFactory delegate = new InvocationHandlerFactory.Default();
+
+    @Override
+    public InvocationHandler create(Target target, Map<Method, MethodHandler> dispatch) {
```
Comment (jamesmartinpp) [inline mocca-client/src/main/java/com/paypal/mocca/client/MoccaFeignInvocationHandlerFactory.java:22]: Does this handle encode/decode exceptions too or only invocation exceptions?

Comment (jamesmartinpp) [reply mocca-client/src/main/java/com/paypal/mocca/client/MoccaFeignInvocationHandlerFactory.java:22]: NVM. I see from the tests below that it does.

**Review by jamesmartinpp — APPROVED** [review]
None of my comments are blocking. This is really cool!

File: mocca-client/src/main/java/com/paypal/mocca/client/MoccaFeignEncoder.java (hunk 3 above)
Comment (fabiocarvalho777) [reply mocca-client/src/main/java/com/paypal/mocca/client/MoccaFeignEncoder.java:102]: There will still be a message. There is always the default message, which has to the field that failed validation and the validation failure. There is also an opportunity for the application owner to customize the messages when they add Bean Validations annotations.

File: build.gradle (hunk 1 above)
Comment (fabiocarvalho777) [reply build.gradle]: Sure

Comment (fabiocarvalho777) [reply build.gradle]: Done

File: docs/END_USER_DOCUMENT.md [hunk 5]
```diff
@@ -316,8 +316,9 @@ In the example above, when calling `addAuthor`, Mocca will not include the list
 For a slightly more complicated example, let's say you wanted to add a book along with the book's author but don't want to include the list
 of books written by the author. You would use the `ignore` property like this:
 
-```java
+``` java
```
Comment (abhijith-prabhakar) [inline docs/END_USER_DOCUMENT.md:319]: adding space will take out the formatting I believe

File: mocca-client/src/main/java/com/paypal/mocca/client/MoccaExceptionHandler.java [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+import javax.validation.ConstraintViolationException;
+import java.util.HashSet;
+import java.util.Set;
+
+/**
+ * This exception handler allows Mocca to hide Feign exception types from applications
+ *
+ * @author facarvalho
+ */
+class MoccaExceptionHandler<T, R> {
```
Comment (abhijith-prabhakar) [inline mocca-client/src/main/java/com/paypal/mocca/client/MoccaExceptionHandler.java]: should be marked final with private constructor.

File: mocca-client/src/main/java/com/paypal/mocca/client/MoccaExceptionHandler.java [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+
+    static {
+        acceptableExceptions.add(MoccaException.class);
+        acceptableExceptions.add(ConstraintViolationException.class);
+    }
+
+    static RuntimeException handleException(Throwable throwable) {
+        if (throwable instanceof FeignException) {
+            Throwable cause = throwable.getCause();
+            if (acceptableExceptions.contains(cause.getClass()) && cause instanceof RuntimeException) {
```
Comment (abhijith-prabhakar) [inline mocca-client/src/main/java/com/paypal/mocca/client/MoccaExceptionHandler.java]: 2nd check is redundant...right? all acceptableExceptions are Runtime exceptions

File: mocca-client/src/main/java/com/paypal/mocca/client/MoccaFeignInvocationHandlerFactory.java [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+import feign.InvocationHandlerFactory;
+import feign.Target;
+
+import java.lang.reflect.InvocationHandler;
+import java.lang.reflect.Method;
+import java.util.Map;
+
+/**
+ * This class allows Mocca to hide Feign exception types from applications
+ * when using a custom Feign invocation handler factory is possible
```
Comment (abhijith-prabhakar) [inline mocca-client/src/main/java/com/paypal/mocca/client/MoccaFeignInvocationHandlerFactory.java:13]: nice and thank you for looking at this

**Review by abhijith-prabhakar — APPROVED** [review]
Couple of minor comments.  PR itself looks good.

File: docs/END_USER_DOCUMENT.md (hunk 5 above)
Comment (fabiocarvalho777) [reply docs/END_USER_DOCUMENT.md:319]: It is actually the other way around. The lack of space breaks formatting.

File: mocca-client/src/main/java/com/paypal/mocca/client/MoccaExceptionHandler.java (hunk 6 above)
Comment (fabiocarvalho777) [reply mocca-client/src/main/java/com/paypal/mocca/client/MoccaExceptionHandler.java]: Done

File: mocca-client/src/main/java/com/paypal/mocca/client/MoccaExceptionHandler.java (hunk 7 above)
Comment (fabiocarvalho777) [reply mocca-client/src/main/java/com/paypal/mocca/client/MoccaExceptionHandler.java]: Done

File: mocca-client/src/main/java/com/paypal/mocca/client/MoccaExceptionHandler.java [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+    static {
+        acceptableExceptions.add(MoccaException.class);
+        acceptableExceptions.add(ConstraintViolationException.class);
+    }
+
+    static RuntimeException handleException(Throwable throwable) {
+        if (throwable instanceof FeignException) {
+            Throwable cause = throwable.getCause();
+            if (acceptableExceptions.contains(cause.getClass())) {
+                return (RuntimeException) cause;
```
Comment (crankydillo) [inline mocca-client/src/main/java/com/paypal/mocca/client/MoccaExceptionHandler.java]: I have these concerns:

1. I don't think we should wrap a `Error` with a `RuntimeException`.
2. I think this was done due to some doc stuff around how a user would get to constraint violations, which come via exceptions.  However, there are other checked exception scenarios that make sense to write code against.  The most notable being IO exceptions due to the HTTP call.  Users will still have to go through feign for that.
3. Feign can have bugs with code in their exception wrapping.  If you rip the cause (the wrapped exception) out, then we lose that stack trace for debugging purposes.  
4. I find it a little weird that we re-throw some Feign runtime exceptions as raw, but wrap other runtime exceptions.  But I guess this is inline with the general vibe, which is to make our doc correct (but this seems like a weird-ish motivator to me).

That said, I don't have huge problems with this.  We can let customers complain (although minimally I would not wrap `Error`).

Comment (fabiocarvalho777) [reply mocca-client/src/main/java/com/paypal/mocca/client/MoccaExceptionHandler.java]: 1. **Concern 1**: Agree. Addressed in my most recent commit. Please take a look at it, I have just pushed it. Thanks for catching that.
2. **Concern 2**: I think that can be a potential enhancement for the future, once we understand that better, especially with feedback from users. For now we are handling the exception types and use cases we are aware of.
3. **Concern 3**: that is not a problem. And here is why. We are being very careful and selective with the exception types we choose to expose (discarding the Feign exception). Right now there are only two types and we are 100% certain that it is totally fine to discard the Feign exception in those cases (`MoccaException` itself and `ConstraintViolationException`). **Important**: notice we are NOT always discarding the Feign exception. We only do that for those two cases. All other cases we preserve it (but we wrap it in `MoccaException`). So I think we are good here.
4. **Concern 4**: I am not sure I understood what you said here, but I believe my comments above might explain it (especially after my most recent commit).

**Review by AmandeepSingh97 — APPROVED** [review]
LGTM

File: mocca-client/src/main/java/com/paypal/mocca/client/MoccaExceptionHandler.java [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+    }
+
+    private static final Set<Class<? extends Throwable>> acceptableExceptions = new HashSet<>();
+
+    static {
+        acceptableExceptions.add(MoccaException.class);
+        acceptableExceptions.add(ConstraintViolationException.class);
+    }
+
+    static Throwable handleException(Throwable throwable) {
```
Comment (crankydillo) [inline mocca-client/src/main/java/com/paypal/mocca/client/MoccaExceptionHandler.java:39]: IMO, we should unit test for this.  I don't know that our functional testing will be sufficient to uncover some cases we'd like to cover (i.e. not wrapping Error).  But, I know there are different stances on what 'unit testing' looks like.  I leave it up to you:)

Comment (fabiocarvalho777) [reply mocca-client/src/main/java/com/paypal/mocca/client/MoccaExceptionHandler.java:39]: Added unit test
